teps/0090-matrix.md

```diff
@@ -596,17 +660,1110 @@ For example:
+
+#### Combinations of Parameters in the Matrix
+
+We need a way to identify a specific combination of `Parameters` in the `Matrix` so that
+we can map it to a given `TaskRun` and its components, such as `Results`. We cannot rely
+on the ordering of the `Parameters` in the `Matrix` to be maintained as specified in the
+`Pipeline`. We propose that we add the generated combinations identifications to the
+`PipelineRunStatusFields` in `PipelineRunStatus`. For example:
+
+```yaml
+  taskRuns:
+    matrixed-pr-foo-0:
+      pipelineTaskName: foo
+      matrixId: 0
+      status:
+        ...
+    matrixed-pr-foo-1:
+      pipelineTaskName: foo
+      matrixId: 1
+      status:
+        ...
+```
+
+The `Parameters` were used for a given `TaskRun` or `Runs` would available in the `TaskRun`
+status itself. The `matrixId` in the `PipelineRunStatus` is useful in identifying that the
+`TaskRun` is part of a `Matrix`, especially if the `TaskRun` name was too long so got hashed.
+This is also in line with providing the minimal information needed in the `PipelineRunStatus`
+as discussed in [TEP-0100: Embedded TaskRuns and Runs Status in PipelineRuns][tep-0100].
+
+### Results
+
+#### Specifying Results in the Matrix
+
+`Results` from previous `TaskRuns` or `Runs` can be passed into the `Matrix`, which will
+dynamically generate `TaskRuns` or `Runs` from the fanned out `PipelineTask`. Today, we
+support string `Results` only, so they will be passed individually into the `Matrix`:
+
+```yaml
+tasks:
+...
+- name: task-4
+  taskRef:
+    name: task-4
+  matrix:
+  - name: values
+    value: 
+    - (tasks.task-1.results.output) # string
+    - (tasks.task-2.results.output) # string
+    - (tasks.task-3.results.output) # string
+```
+
+When we support array `Results`, as proposed in [TEP-0076][tep-0076], users can pass in
+array `Results` directly into the `Matrix`:
+
+```yaml
+tasks:
+...
+- name: task-5
+  taskRef:
+    name: task-5
+  matrix:
+  - name: values
+    value: (tasks.task-4.results.output) # array
+```
+
+#### Results from Fanned Out PipelineTasks
+
+Today, each `PipelineTask` is executed in a single `TaskRun` or `Run` which produces the
+`Results` specified in the underlying `Task` or `Custom Task`. Each `Result` is accessible
+through a variable - `$(tasks.<pipelinetask-name>.results.<result-name>)`.
+
+With a `Matrix`, the `PipelineTask` will be executed in several `TaskRuns` or `Runs`, each
+of which will produce its own `Results`. Therefore, we need to adjust the data structure for
+`Results` when a `Matrix` is used to fan out to accommodate all the `Results` produced.
+
+When the underlying `Task` produces a `Result`, we propose that the fanned out `PipelineTask`
+produces a `Result` of type `Object` which maps a specific combination identification to the
+`Result` from that combination's `TaskRun` or `Run`. The `Result` will be accessible via the
+same variable as is right now - `$(tasks.<pipelinetask-name>.results.<result-name>)`.
```

Comment:
Following up on our discussion Wednesday, I agree with @wlynch 's hesitation around changing the structure of the result based on how it is used

(also apologies @jerop b/c we definitely discussed this ages ago as well and now im changing my mind XD)

i.e. when used in a non-matrix-ed pipline task,  `$(tasks.<pipelinetask-name>.results.<result-name>)` would have the type declared by the task `<pipelinetask-name>` is using, but when used with a matrix, this is now always an array, where each individual value is the type declared by the task

the alternatives section explores a couple of other options but mentions challenges due to the dynamic nature of the results (you don't know how many there will be in advance, or what the taskrun names will be, and you still want to be able to map these results back to the individual taskrun that produced them)

@wlynch mentioned an alternative which would involve leaning on json path that i think it might be worth pursuing, as a way to:
* pick a structure that doesn't make the type of the result change based on how it's used
* make it possible to map the results back to the individual taskruns
* not require you to know the number of elements on their names in advance

in order for this to help I think we'd need to make a change to how the results are emitted in this case, i have 2 ideas which both have their own downsides:

1. Replace `<pipelinetask-name>` with the name of the matrixed taskrun, i.e. `$(tasks.<matrixed-taskrun>.results.<result-name>)`, e.g. `$(tasks.test-chrome-linux.results.<result-name>)`
2. When running in matrix mode, the value at `tasks.<pipelinetask-name>` becomes a map of the matrixed item (e.g. taskrun name or just identifiers like `chrome-linux`) to the per. taskrun variable replacements, i.e. `$(tasks.<pipelinetask-name>.<matrixed-taskrun>.results.<result-name>)`, e.g. `$(tasks.test.chrome-linux.results.<result-name>)`
2b. Same as 2, but nest the results for each combo, e.g. `$(tasks.test.chrome.linux.results.<result-name>)`

(1) is odd b/c it means that sometimes the keys in `tasks` are the pipeline task names, and sometimes the taskrun names, im not a huge fan of the inconsistency
(2) still changes the shape of the variable replacement, but I like that it doesn't change the shape of the result itself (i.e. a string result is still a string)

"but wait," i hear you say "this doesn't help any of the problems I already identified! you won't know `chrome-linux` or the taskrun name or the number of items, etc. in advance!"

this is where the jsonpath option makes things interesting! in the current proposal, conceptually we are providing variable replacement with a json object like this:

```json
{
  "tasks": {
    "test": {
      "results": {
        "output": {
          "chrome": "45 tests passed",
          "firefox": "44 tests passed, 1 failed"
        }
      }
    }
  }
}
```

let's say we go with option 2 (`$(tasks.<pipelinetask-name>.<matrixed-taskrun>.results.<result-name>)`), then we'd be providing a json object like this:

```json
{
  "tasks": {
    "test": {
      "chrome": {
        "results": {
          "output": "45 tests passed"
        }
      },
      "firefox": {
        "results": {
          "output": "44 tests passed, 1 failed"
        }
      }
    }
  }
}
```

So you could access a specific string value with `$(tasks.test.chrome.results.output)` which we already know BUT what if we also allowed `$(tasks.test)`? which could resolve to:

```json
[
  {
    "chrome": {
      "results": {
        "output": "45 tests passed"
      }
    },
    "firefox": {
      "results": {
        "output": "44 tests passed, 1 failed"
      }
    }
  }
]
```

And if we later let you use more json path syntax, you could change the structure if you wanted, e.g.:

```
$(tasks.test[*].results.output)
```
Would result in:
```json
[
  "45 tests passed",
  "44 tests passed, 1 failed"
]
```

Halfway through this I realized it wasn't really the additional json path features part that was important, it was really us shifting the syntax a little to treat the variable replacement as json path being applied to a json object -a and expanding that object a little for this use case

Reply:
Thanks for this detailed follow up Christie! 

I like this suggestion, and would love to explore it more when TEP-0075 and TEP-0075 have landed. The design mentioned was just one option, but as noted much further down solving for this is not in scope for this TEP. Removed this section from the proposal, and added an explicit non-goal. Excited to explore this further later!